File: 프로그래머스/3/92343. 양과 늑대/양과 늑대.py
from collections import defaultdict

def solution(info, edges):
    global answer
    answer = 0
    tree = defaultdict(list)
    for u,v in edges :
        tree[u].append(v)
    
    def dfs(node, sheep, wolf, possible):
        global answer
        
        # 현재 양과 늑대 상태가 유효하지 않으면 중단
        if sheep <= wolf:
            return
    
        answer = max(answer, sheep)
        
        for nxt in possible:
            new_possible = possible.copy()
            new_possible.remove(nxt)
            new_possible.extend(tree[nxt])
            
            if info[nxt] == 0:
                dfs(nxt, sheep + 1, wolf, new_possible)
            else:
                dfs(nxt, sheep, wolf + 1, new_possible)
    
    dfs(0, 1, 0, tree[0])
    return answer

File: 프로그래머스/3/92343. 양과 늑대/양과 늑대.java
class Solution {
    public int solution(int[] info, int[][] edges) {
        int n = info.length;
        boolean[] visited = new boolean[n];
        visited[0] = true;
        int answer = dfs(info, edges, 1,0, visited);
        return answer;
    }
    
    // 부모 방문 & 자식 방문하지 않은 경우에 탐색
    int dfs(int[] info,int[][] edges,int sheep, int wolf, boolean[] visited){
        if (sheep == wolf) return sheep;
        int maxSheep = sheep;
        
        for (int[] edge : edges){
            int parent = edge[0];
            int child = edge[1];
            if (visited[parent] && !visited[child]){
                visited[child] = true;
                if (info[child] == 0){
                    maxSheep = Math.max(maxSheep, dfs(info, edges, sheep + 1, wolf,visited)) ;    
                }else {
                    maxSheep = Math.max(maxSheep, dfs(info, edges, sheep, wolf + 1,visited)) ;
                }
            visited[child] = false;                 
            }

        }
        return maxSheep;
    }
}